### ISSUE TITLE
Trigger not invoked when SaveChanges is called from exception handler

### ISSUE THREAD

**Issue by Fulgan** [koenbeuk/EntityFrameworkCore.Triggered#110]
I have the following situation:

My application implements a rather long and complex import process. Since it takes too much time to wrap it into a transaction, I have to keep track of created objects and perform cleanup in case of an exception being raised.

In parallel, I have added a trigger to ensure that a specific (soft) constraint is respected (it's actually checking that the operation doesn't cause the number count of one specific entity type doesn't go over the lincesed limit). When that triggers determines that the operation needs to be aborted, it raises a specialized exception.

My problem is that it seems that while that trigger works fine elsewhere in the application, it fails to prevent the entity creating in this import process.

Here is the pseudo-code for the issue:

```CSharp
var entity = new Entity();
dbContext.Add(entity);
try
{
  dbContext.SaveChanges(); // <-- this will cause the trigger to raise an exception
} catch {
  // Do some necessary cleanup
  //...
  
  dbContext.SaveChanges(); // <-- At this stage, the entity is still in the "added" state in the change tracker. The trigger doesn't get called anymore and therefore the entity is added.
}

```
How can I enforce the trigger to be called regardless of wether it has been called already before ? Is there an alternate way to ensure the second call does not add the entity to the model again? I have tried changing te state to "untracked" but that didn't help.

**Comment by koenbeuk** [comment]
By default we indeed keep track of what triggers have been raised so that we don't 'double' raise them which is typically the desired behavior. 

Your use-case is fair though, What you would need is a manual way to reset the TriggerSession, we can ALMOST do that today:
```csharp
var triggerService = dbContext.GetService<ITriggerService>();
var triggerSession = triggerService.Current; // This will be null when we've not yet called SaveChanges before
triggerSession.Reset(); // This proposed method should reset the tracked trigger invocations
```
This should not be too difficult to implement.

Meanwhile there is a way of getting the desired behavior by actually relying on a known bug. With EF Core 3.1, we had to derive our DbContext from `TriggeredDbContext` (Since SaveChangesInterception was not yet a thing in EF). Doing so today will cause Triggers to be raised by the `TriggeredDbContext` base class instead of SaveChangesInterception. The TriggeredDbContext does not share state between multiple calls to SaveChanges so that each subsequent call to SaveChanges will indeed re-raise all triggers.

In short, change: `class MyDbContext : DbContext` into `class MyDbContext : TriggeredDbContext` and your desired use-case should work. Meanwhile, I'll plan on implementing a Reset option for your trigger session so that you can use that instead of this work around.

**Comment by koenbeuk** [comment]
Perhaps I was misinterpreting your issue. If you want all triggers to be raised again on subsequent calls to SaveChanges then my above answer is relevant. If you want to make sure that the entity does NOT get included in the Save set then you can control that from within your trigger:

```csharp
public class RateLimitTrigger : IBeforeSaveTrigger<MyEntity>, IBeforeSaveCompletedTrigger {
   int _entityCount;

   public RateLimitTrigger(MyDbContext dbContext) { ... }
    
   public Task BeforeSave(ITriggerContext<MyEntity> context, CancellationToken) {
       _entityCount += 1;
       if (_entityCount > MaxNumberOfEntitiesAllowed) { 
          // Unattach the entity from my DbContext, this way it wont be included
          context.ChangeTracker.Entry(context.Entity).State = State.Detached;
          
          // Raise an exception here if you want to abort early. Keep in mind that this will prevent subsequent triggers to be invoked
          // There may be additional Entities that have been added for which this trigger should still be called 
          // throw new MySpecialException("Rate limit reached...");
       }
   }
   
   public Task BeforeSaveCompleted(CancellationToken) {
      // Using a IBeforeSaveCompleted trigger allows all the RateLimitTrigger to call BeforeSave on all entities
      // thereby ensuring that all 'disallowed' entities have been Detached from the database.
      if (_entityCount > MaxNumberOfEntitiesAllowed) { 
          throw new MySpecialException("Rate limit reached...");
      }
   }   
}
```
Finally, we used State in our trigger (namely we kept track of _entityCount), So we need to register our trigger with a Scoped lifetime.

```csharp
triggerOptions.AddTrigger<RateLimitTrigger>(ServiceLifetime.Scoped);
```
Hope this or the above answer helps

**Comment by Fulgan** [comment]
Thank you for your answer.

I think the best would be to be able to controle whether or not the trigger can be re-raised or not from inside the trigger itself: you've given me a couple of way to verify that. 

I also realize that my original code was also bugged: I need to treat that entity as part of the ones that needs to be cleand up.

Again, tahnk you.

**Comment by koenbeuk** [comment]
@Fulgan Please let me know how the cleanup went and if you were able to solve your problems with help of my answers above. I'm not sure if we still have a use-case for being able to re-raise triggers. Let me know if you think otherwise.

**Comment by Fulgan** [comment]
@koenbeuk I was able to solve my problem by changing cleanup code to take into account the fact that the trigger doesn't prevent the object's creation and manually deleting it after the fact.

It's not ideal since it means that the code has to have to know about these triggers explicitely in order to perform the cleanup and therefore means that adding new triggers to do the same kind of job might impact old code.

Overally, I'm still wondering about the wisedome of not calling the triggers multiple times if necessary: the more I think about it, the more it makes me feel like this can only cause problems. After all, if the trigger is raised a second time in the same context it's that the entity it is acting upon has changed somehow. And since the trigger is only being called when the entity is being persisted, it seems to me that the general case would be that the trigger should, be default, be called multiple times and that the case where subsequent calls should be supressed is the edge case here.

**Comment by koenbeuk** [comment]
> if the trigger is raised a second time in the same context it's that the entity it is acting upon has changed somehow

This would be true under a full cascading strategy where each change to an Entity would re-raise triggers. This strategy is not implemented as I've not yet seen a valid use case. EntityAndType (default) where we only raise Triggers for unique Entity/Change type combinations has however very practical use cases. Lets take a look at this [Student Manager sample](https://github.com/koenbeuk/EntityFrameworkCore.Triggered/tree/master/samples/v2/3%20-%20StudentManagerAspNetCore) (Part of this Repo, you can find more about these triggers by checking the sample and its readme):

- **Courses.AutoSignupStudents**: Breaks if this is ran for the same student multiple times (could be fixed with some additional checks)
- **StudentCourses.BlockRemovalWhenCourseIsMandatory:** No need to run twice
- **StudentCourses.SendWelcomingEmail:** This should always just run once! It may also not be trivial to implement a check to ensure that it did not run before
- **Students.AssigRegistrationDate:** No need to run twice, could cause additional side effects
- **Students.SignupToMandatoryCourses:** Breaks if this is ran multiple times
- **Traits.Audited.CreateAuditRecord:** Breaks if this is ran multiple times
- **Traits.SoftDelete.EnsureSoftDelete:** No need to run twice

In this sample project, none of the triggers would benefit from being able to be called multiple times for the same Entity / Change type.

I feel like I don't completely understand the problem you're trying to solve and thereby the solution you came up with (and therefore why it would benefit you if the Trigger was called multiple times for the same Entity / Change type). Can you perhaps share some sample code of what you've implemented right now?

**Comment by Fulgan** [comment]
The specific use case I have is to enforce a domain constraint that cannot be enforced through a regular SQL constraint.

My real use case if the following: I'm writing a multi-tenant accounting solution. Data in a given tenant is grouped by "Company" each containing a number of dependant "FiscalYear" (etc.). I have a requirement to limit the number of companies allowed per tenant based on the licensing configuration.

Making sure that the user is always prevented from creating a new company is hard because there are many code path that allows them to do so: imported from one of the several supported formats, restoring from a backup, copy from anyother Company entity, etc.

I could solve that using a repository pattern but doing so when using EF is problematic: the repository pattern is actually doing the same job as the ORM but with less flexibility. What it does, however, is allow me to define domain-level constraints that I cannot express easily with EF.

Using a trigger allows me to ensure that these domain constraints will be enforced in all cases even if the programmer who implemented the functionality didn't verify it manually without requiring an added abstraction layer.

I could implement these domain constraints in my own SaveChangesInterceptor (Audit tracing is already implemented this way ) but I would rather use a more generic system and implementing one pretty much means rewriting triggers in a similar way as your lib.

**Comment by koenbeuk** [comment]
Yes Triggers are an excellent candidate for this, have you tried the solution proposed [here?](https://github.com/koenbeuk/EntityFrameworkCore.Triggered/issues/110#issuecomment-858541120). It may require some adaptation but it sounds like this is what you are looking for. 

If not, can you clarify what is lacking and why you need to re-raise triggers? Perhaps you can prototype out what your own custom SaveChangesInterceptor  could look like? Thanks!

**Comment by Fulgan** [comment]
Hello,

Sorry but it took me a while to confirm that your proposed sollution was indeed working for us: it does. Thank you again.
